### PR TITLE
SLCORE-999 Store custom impacts when synchronizing rules

### PR DESCRIPTION
**Pull request by eray-felek-sonarsource**
[SLCORE-999](https://sonarsource.atlassian.net/browse/SLCORE-999)

[SLCORE-999]: https://sonarsource.atlassian.net/browse/SLCORE-999?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java [hunk 1]
```diff
@@ -230,19 +230,17 @@ private ServerActiveRule tryConvertDeprecatedKeys(ServerActiveRule possiblyDepre
       var ruleKeyPossiblyWithDeprecatedRepo = RuleKey.parse(possiblyDeprecatedActiveRuleFromStorage.getRuleKey());
       var templateRuleKeyWithCorrectRepo = RuleKey.parse(ruleOrTemplateDefinition.get().getKey());
       var ruleKey = new RuleKey(templateRuleKeyWithCorrectRepo.repository(), ruleKeyPossiblyWithDeprecatedRepo.rule()).toString();
-      //TODO Replace empty list with proper one.
       return new ServerActiveRule(ruleKey, possiblyDeprecatedActiveRuleFromStorage.getSeverity(), possiblyDeprecatedActiveRuleFromStorage.getParams(),
-        ruleOrTemplateDefinition.get().getKey(), Collections.emptyList());
+        ruleOrTemplateDefinition.get().getKey(), possiblyDeprecatedActiveRuleFromStorage.getOverriddenImpacts());
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java:234]: Same here

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java [hunk 2]
```diff
@@ -230,19 +230,17 @@ private ServerActiveRule tryConvertDeprecatedKeys(ServerActiveRule possiblyDepre
...
       ruleOrTemplateDefinition = rulesRepository.getRule(connectionId, possiblyDeprecatedActiveRuleFromStorage.getRuleKey());
       if (ruleOrTemplateDefinition.isEmpty()) {
         // The rule is not known among our loaded analyzers, so return it untouched, to let calling code take appropriate decision
         return possiblyDeprecatedActiveRuleFromStorage;
       }
-      //TODO Replace empty list with proper one.
       return new ServerActiveRule(ruleOrTemplateDefinition.get().getKey(), possiblyDeprecatedActiveRuleFromStorage.getSeverity(),
         possiblyDeprecatedActiveRuleFromStorage.getParams(),
-        null, Collections.emptyList());
+        null, possiblyDeprecatedActiveRuleFromStorage.getOverriddenImpacts());
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java:243]: Same here

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java [hunk 3]
```diff
@@ -230,19 +230,17 @@ private ServerActiveRule tryConvertDeprecatedKeys(ServerActiveRule possiblyDepre
...
     }
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java:244]: This method is very similar to the one in AnalysisService, maybe there is a way to keep only one

Comment (eray-felek-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RulesService.java:244]: How about making it static on RuleService and using it on AnalysisService?

**Review by damien-urruty-sonarsource — APPROVED** [review]
Some improvement ideas, but overall LGTM

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RuleDetails.java [hunk 4]
```diff
@@ -147,11 +149,25 @@ public static RuleDetails merging(ServerActiveRule activeRuleFromStorage, Server
       serverRule.getSeverity(),
       templateRuleDefFromPlugin.getType(),
       cleanCodeAttribute,
-      defaultImpacts,
+      mergeImpacts(defaultImpacts, activeRuleFromStorage.getOverriddenImpacts()),
```
[the remark this answers is not in the document]
Comment (eray-felek-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/rules/RuleDetails.java:152]: I agree in can be improvement for later. Created this ticket to track: https://sonarsource.atlassian.net/browse/SLCORE-1021

**Pull request merged**